docs/BestPractices.md

```diff
@@ -0,0 +1,123 @@
+- *Don’t*: Delete outputs from steps that passed when the full pipeline fails, keeping them enables you to pick up where you left off.
+- *Don’t*: Use tools that are “buggy” or fragile, find alternatives or improve the tools.
+  
+## Testability
+### What is a testable pipeline?
+A testable pipeline is one in which isolated sections or the full pipeline can checked for specified characteristics without modifying the pipeline’s code. Testability requires the existence of appropriate data with which to run the test and a testing checklist that reflects a clear understanding of how the data will be used to evaluate the pipeline.
+  
+### Why do we care about testabilty?
+The availability of test data enables validation that the pipeline can produce the desired outcome. Formulation of a testing checklist allows the developer to clearly define the capabilities of the pipeline and the parameters of its use.
+
+- *Do*: Provide example test data with your pipeline/tool.
+- *Do*: Provide the results of an execution of your pipeline/tool on the test data.
+- *Do*: Refer to at least one real data set appropriate for your tool/pipeline with example output from an execution of your pipeline or tool.
+- *Do*: Provide a checker tool.
+- *Do*: Include an automated testing suite for pipeline tasks.
+- *Do*: Include automated tests for the pipeline as an integrated unit (pipeline benchmarking).
+- *Don’t*: Use tools that do not have automated testing suites.
+- *Don’t*: Write tests that assume a specific instance of data.
+
+## Portability
+
+### What is pipeline portability?
+Pipeline portability refers to the ability of a pipeline to execute successfully on multiple technical architectures.
+
+### Why do we care about portability?
+_Science._ Science is not science if results are not reproducible; the scientific method cannot occur without a repeatable experiment that can be modified. Data processing pipelines are an essential part of some scientific inquiry and where they are leveraged they should be repeatable to validate and extend scientific discovery.
+
+_Impact._ Pipelines will have greatest impact when they can be leveraged in multiple environments. The more technical requirements for installing and running of a pipeline, the longer it will take for a researcher to have a usable running pipeline.
+  
+_Maintainability._ Over the long term, it is easier to maintain pipelines that can be run in multiple environments. Portability avoids being tied to specific infrastructure and enables ease of deployment to development environments.
+
+To ensure that others will be able to use your pipeline, avoid building in assumptions about environments and infrastructures in which it will run. 
+
+### Configurability for running on different technical infrastructures.
+Code should not change to enable a pipeline to run on a different technical architecture; this change in execution environment should be configurable outside of the pipeline code.
```

Comment:
This sounds great, but I'm not sure it is entirely true for us at this point. We still have infrastructure specific code in the runtime settings for our tasks. As long as these are just best practices and not what we promise we are doing, then this seems good